### PR TITLE
support storing detector result to customr result index

### PR DESCRIPTION
**Pull request by ylwu-amzn**
Signed-off-by: Yaliang Wu <ylwu@amazon.com>

### Description

Now we store anomaly result in system index. User can user search anomaly result API to query data in result system index, but user can't search that index to build their own dashboard. This PR add option to let user specify their own custom result index and AD plugin will store detector's result to that index. So user can get detector result by searching that index directly.

Main changes:
1. Add result index option on detector creation page
2. Show result index on detector configuration review page
3. Read anomaly result from custom result index on detector realtime/historical result page.
4. Read anomaly result from custom result indices on AD dashboard and detector list page.

TODO: fix some test ; add more test for new code

Screen Shot 2021-10-29 at 11 42 55 AM

Screen Shot 2021-10-28 at 4 17 11 PM

### Check List

- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
- [ ] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx
Comment (ohltyler) [inline public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx:173]: nit - you can move to the top of the file so it only needs to be declared once

**Comment by ohltyler** [comment]
Couple high-level things:
1. How does a user default to not using a custom result index, should there be a checkbox or something to select/de-select?
2. What shows on the detector settings page if no custom index was selected?

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (ohltyler) [inline public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx:106]: is there any restrictions on the backend for this? If so can that be added as help text underneath this text form?

File: public/pages/DetectorResults/containers/AnomalyHistory.tsx
Comment (ohltyler) [inline public/pages/DetectorResults/containers/AnomalyHistory.tsx:178]: same thing here - maybe move to the top of the file so it only needs to be declared once

File: public/pages/DetectorResults/containers/AnomalyHistory.tsx
Comment (ohltyler) [inline public/pages/DetectorResults/containers/AnomalyHistory.tsx:354]: most of the queries / results parsing has been changed as part of multi-category filtering PR. Will need to keep an eye on that when merging.

File: public/redux/reducers/anomalyResults.ts
Comment (ohltyler) [inline public/redux/reducers/anomalyResults.ts:109]: can remove this comment

File: public/utils/constants.ts
Comment (ohltyler) [inline public/utils/constants.ts:71]: `LINE` -> `LINK`

File: server/cluster/ad/adPlugin.ts
Comment (ohltyler) [inline server/cluster/ad/adPlugin.ts:109]: is it ok that this has the same name as the one implemented directly below? Usual strategy is to name to something more verbose - for example, the start detector has 2 different calls, one for real-time (`ad.startDetector`) and one for historical (`ad.startHistoricalDetector`)

File: server/routes/ad.ts
Comment (ohltyler) [inline server/routes/ad.ts:483]: when would this case occur?

File: server/routes/ad.ts
Comment (ohltyler) [inline server/routes/ad.ts:485]: by specifying the string as `''`, does this indicate no custom result index is specified? If so can you add this in comments?

File: server/routes/ad.ts
Comment (ohltyler) [inline server/routes/ad.ts:600]: can you explain how this includes results from custom result indices and non-custom results indices?

File: server/routes/ad.ts
Comment (ohltyler) [inline server/routes/ad.ts:755]: same thing here, maybe add some comment

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (ohltyler) [inline public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx:102]: Seems weird to see this text when editing and it's read-only. How about change to empty if `props.isEdit` is `true`?

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (ylwu-amzn) [reply public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx:106]: Backend only restrict the prefix, for other limitation will leave to OpenSearch core to detect

File: server/routes/ad.ts
Comment (ohltyler) [reply server/routes/ad.ts:483]: after discussion, just an extra check for some adversarial request.

File: server/routes/ad.ts
Comment (ohltyler) [reply server/routes/ad.ts:600]: After discussion, understood that the backend will search on the system indices by default, + results from any specified custom result index. So by specifying all custom result indices, all results (from both system indices + any custom results indices) will be returned here.

Can you add a comment in the code about this @ylwu-amzn ?

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (ylwu-amzn) [reply public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx]: Refactored the whole custom result index component based on CX designer's suggestion. Just use similar design or category field. See the latest screenshot.

File: server/routes/ad.ts
Comment (ylwu-amzn) [reply server/routes/ad.ts:760]: done

File: server/routes/ad.ts
Comment (ylwu-amzn) [reply server/routes/ad.ts:604]: Sure, done.

File: server/routes/ad.ts
Comment (ylwu-amzn) [reply server/routes/ad.ts:486]: done.

File: server/cluster/ad/adPlugin.ts
Comment (ylwu-amzn) [reply server/cluster/ad/adPlugin.ts:109]: Actually we don't need this one, as the result index is not required, check line 115 `required: false,`. 
Removed the second `ad.searchResults` and tested, everything looks good.

File: public/utils/constants.ts
Comment (ylwu-amzn) [reply public/utils/constants.ts]: Discussed with UX designer, will remove this link and show AD doc link. Will add more details on AD doc to explain user should use ISM plugin to manage custom result index.

File: public/redux/reducers/anomalyResults.ts
Comment (ylwu-amzn) [reply public/redux/reducers/anomalyResults.ts]: done

File: public/pages/ReviewAndCreate/components/DetectorDefinitionFields/DetectorDefinitionFields.tsx
Comment (ylwu-amzn) [reply public/pages/ReviewAndCreate/components/DetectorDefinitionFields/DetectorDefinitionFields.tsx]: Discussed with UX designer and refactored this part. Will use "Custom result index" and show "-" if no custom result index specified.

File: public/pages/DetectorResults/containers/AnomalyHistory.tsx
Comment (ylwu-amzn) [reply public/pages/DetectorResults/containers/AnomalyHistory.tsx]: Done

File: public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx
Comment (ylwu-amzn) [reply public/pages/AnomalyCharts/containers/AnomalyDetailsChart.tsx]: done

**Comment by ylwu-amzn** [comment]
New custom result index UI
Screen Shot 2021-10-30 at 9 17 05 PM

**Comment by ohltyler** [comment]
Another minor question, maybe it's a detail in the backend PR, need to look more. But what would happen if a user starts a fresh cluster and only creates custom result indices detectors first? Does the default result system index still get inited? And if not, does either frontend or backend handle any errors properly if that index is missing?

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (ohltyler) [inline public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx:57]: nit: `result` -> `results` sounds more grammatically correct

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (ohltyler) [inline public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx:127]: If enabled and `isEdit` is true, shouldn't the selected custom result index still show? Maybe the `!props.isEdit` condition can be removed (current conditions in `prepend` and `disabled` have some unreachable scenarios as well because of `!props.isEdit` condition).

**Comment by ohltyler** [comment]
May need to fix some broken tests and update snapshots with `yarn test:jest <files-to-test> -u`. I'm ok with adding tests as a separate PR - if so, can you create an issue to track?

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (ylwu-amzn) [reply public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx:127]: Good catch, change to show result index when edit detector and remove the callout of `"You can't change the custom result index after you create the detector."` as it's duplicate as the callout message `title="You can't change the custom result index after you create the detector. You can manage the result index with Index Management plugin."`. Will ask UX designer to help review.

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (ylwu-amzn) [reply public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx:57]: will fix

**Comment by ylwu-amzn** [comment]
> Another minor question, maybe it's a detail in the backend PR, need to look more. But what would happen if a user starts a fresh cluster and only creates custom result indices detectors first? Does the default result system index still get inited? And if not, does either frontend or backend handle any errors properly if that index is missing?

Default result index will not be inited if just create detector with custom result index on a fresh cluster.Tested locally, no issue found.

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (amitgalitz) [inline public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx:89]: nit: I believe the second sentence here: "You can manage the result index with Index Management plugin." should be changed to "You can manage the result index with **the** Index Management plugin."

File: public/utils/constants.ts
Comment (amitgalitz) [inline public/utils/constants.ts:82]: nit: type on comment, 'szie' -> 'size'

File: public/utils/utils.tsx
Comment (amitgalitz) [inline public/utils/utils.tsx:72]: In future this method can also be used when user tries inputing custom data source index.

Comment (ylwu-amzn) [reply public/utils/utils.tsx:72]: Currently we allow user to choose concrete index or input index pattern like "abc*". So can't reuse this part directly. We can revisit/refactor this part in future if needed.

File: public/utils/constants.ts
Comment (ylwu-amzn) [reply public/utils/constants.ts:82]: sure, will fix

File: public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx
Comment (ylwu-amzn) [reply public/pages/DefineDetector/components/CustomResultIndex/CustomResultIndex.tsx:89]: sure

**Review by ohltyler — APPROVED** [review]
LGTM!

**Review by amitgalitz — APPROVED** [review]
LGTM, make sure you sign off all commits in the future. Will merge DCO check after this PR is merged so there will be no issue for this PR at the moment.

**Comment by ohltyler** [comment]
Shouldn't the backend have been merged before this one? I guess all will be merged soon enough, so no big issue

**Pull request merged**
